bc_obps/compliance/service/automated_process/compliance_handlers.py

```diff
@@ -0,0 +1,110 @@
+from typing import Protocol
+from compliance.models.elicensing_invoice import ElicensingInvoice
+from compliance.models.compliance_obligation import ComplianceObligation
+from compliance.models.compliance_report_version import ComplianceReportVersion
+from compliance.service.penalty_calculation_service import PenaltyCalculationService
+from compliance.service.compliance_obligation_service import ComplianceObligationService
+from django.utils import timezone
+from decimal import Decimal
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class ComplianceUpdateHandler(Protocol):
+    """Protocol for compliance update handlers."""
+
+    def can_handle(self, invoice: ElicensingInvoice, obligation: ComplianceObligation) -> bool:
+        """Determine if this handler can handle the given invoice and obligation."""
+        ...
+
+    def handle(self, invoice: ElicensingInvoice, obligation: ComplianceObligation) -> None:
+        """Handle the compliance update."""
+        ...
+
+
+class PenaltyPaidHandler(ComplianceUpdateHandler):
+    """Handling invoices with penalties that are fully paid."""
+
+    def can_handle(self, invoice: ElicensingInvoice, obligation: ComplianceObligation) -> bool:
+        """Check if invoice has a penalty and the penalty invoice is fully paid."""
+        has_penalty = hasattr(invoice, 'compliance_penalty') and getattr(invoice, 'compliance_penalty', None)
+        if not has_penalty:
+            return False
+
+        penalty_record = getattr(invoice, 'compliance_penalty', None)
+        if not penalty_record or not penalty_record.elicensing_invoice:
+            return False
+
+        penalty_invoice = penalty_record.elicensing_invoice
+        return penalty_invoice and penalty_invoice.outstanding_balance == Decimal('0.00')
```

Comment:
We can remove the penalty_invoice = penalty_record.elicensing_invoice here. it's the same as the invoice passed as a parameter. Also suggesting a change to the return boolean condition
```suggestion
        return penalty_record.compliance_obligation.penalty_status == 'NOT PAID' and penalty_invoice.outstanding_balance == Decimal('0.00')
```